DeviceAdministration/Infrastructure/Repository/DeviceRegistryRepository.cs

```diff
@@ -108,29 +49,31 @@ public DeviceRegistryRepository(IConfigurationProvider configProvider, IDocDbRes
         /// <returns></returns>
         public async Task<DeviceModel> AddDeviceAsync(DeviceModel device)
         {
-            string deviceId = device.DeviceProperties.DeviceID;
-            DeviceModel existingDevice = await this.GetDeviceAsync(deviceId);
+            if (string.IsNullOrEmpty(device.id))
+            {
+                device.id = Guid.NewGuid().ToString();
+            }
 
+            var id = device.id;
+            DeviceModel existingDevice = await GetDeviceAsync(device.DeviceProperties.DeviceID);
             if (existingDevice != null)
             {
-                throw new DeviceAlreadyRegisteredException(deviceId);
+                throw new DeviceAlreadyRegisteredException(device.DeviceProperties.DeviceID);
             }
 
-            device = (await _docDbRestUtil.SaveNewDocumentAsync<DeviceModel>(device)).ToObject<DeviceModel>();
-
+            await _documentClient.SaveAsync(id, device);
```

Comment:
We were always using id as the key. It was being abstracted away by DocDbRestUtility. It would internally use reflection to look for a id property on the dynamic. The only change I did was to uplevel it. Unncessary usage of reflection to load up a property that we know already exists on the strongly typed model. So explicit passing makes things more transparent and performant.